### PR TITLE
-Werror for libfilament and backend

### PR DESCRIPTION
**Pull request by poweifeng**

### CODE REVIEW

**Review by bejado — APPROVED** [review]

File: filament/backend/src/opengl/platforms/PlatformCocoaTouchGL.mm [hunk 1]
```diff
@@ -88,6 +88,9 @@
             pImpl->mGLContext, nullptr, &pImpl->mTextureCache);
     assert_invariant(success == kCVReturnSuccess);
 
+    // Necessary to avoid unused-variable warning in release build.
+    (void)success;
```
Comment (pixelflinger) [inline filament/backend/src/opengl/platforms/PlatformCocoaTouchGL.mm:92]: you can use `UTILS_UNUSED_IN_RELEASE` instead. e.g.:

```
UTILS_UNUSED_IN_RELEASE bool success;
```

**Review by pixelflinger — CHANGES_REQUESTED** [review]
I don't understand why the changes to `HwRenderPrimitiveFactory`, the comparators were written this way for better performance.

**Comment by poweifeng** [comment]
> I don't understand why the changes to `HwRenderPrimitiveFactory`, the comparators were written this way for better performance.

It's because of this: https://github.com/google/filament/actions/runs/4148557313/jobs/7176658562#step:3:2325

Though I'm not sure if that's a bug in clang-8 or not, I'm unable to repro on my local machine of clang-14.

**Comment by poweifeng** [comment]
> > I don't understand why the changes to `HwRenderPrimitiveFactory`, the comparators were written this way for better performance.
> 
> It's because of this: https://github.com/google/filament/actions/runs/4148557313/jobs/7176658562#step:3:2325
> 
> Though I'm not sure if that's a bug in clang-8 or not, I'm unable to repro on my local machine of clang-14.

I'll dig a little further

**Comment by pixelflinger** [comment]
> > > I don't understand why the changes to `HwRenderPrimitiveFactory`, the comparators were written this way for better performance.
> > 
> > 
> > It's because of this: https://github.com/google/filament/actions/runs/4148557313/jobs/7176658562#step:3:2325
> > Though I'm not sure if that's a bug in clang-8 or not, I'm unable to repro on my local machine of clang-14.
> 
> I'll dig a little further

How did it work until now? Did it generate a warning only on that target?
We had many issues with that specific target. 
Maybe you can just disable that warning.

**Comment by poweifeng** [comment]
> > > > I don't understand why the changes to `HwRenderPrimitiveFactory`, the comparators were written this way for better performance.
> > > 
> > > 
> > > It's because of this: https://github.com/google/filament/actions/runs/4148557313/jobs/7176658562#step:3:2325
> > > Though I'm not sure if that's a bug in clang-8 or not, I'm unable to repro on my local machine of clang-14.
> > 
> > 
> > I'll dig a little further
> 
> How did it work until now? Did it generate a warning only on that target? We had many issues with that specific target. Maybe you can just disable that warning.

It's been throwing a warning for previous builds: https://github.com/google/filament/actions/runs/4167262765/jobs/7212628562#step:3:2395

But it is a bug with clang: https://reviews.llvm.org/D60540

I think we might be able to resolve it if we update clang.

File: filament/backend/src/opengl/platforms/PlatformCocoaTouchGL.mm (hunk 1 above)
Comment (poweifeng) [reply filament/backend/src/opengl/platforms/PlatformCocoaTouchGL.mm]: done

**Review by pixelflinger — APPROVED** [review]

**Pull request merged**